code/modules/fluids/fluid_pipe_machinery.dm

```diff
@@ -0,0 +1,521 @@
+	interact_particle(user, src)
+	if(ON_COOLDOWN(src, "fluidvalve", 1 SECOND))
+		return
+	src.on = !src.on
+	src.UpdateIcon(TRUE)
+	user.visible_message(SPAN_NOTICE("[user] turns [src.on ? "on" : "off"] [src]."), SPAN_NOTICE("You turn [src.on ? "on" : "off"] [src]."))
+	if(!(src.network1 && src.network2))
+		return
+	if(src.on)
+		src.network1.merge_network(src.network2)
+	else
+		src.network1.rebuild_network()
+
+/obj/machinery/fluid_machinery/binary/valve/update_icon(animate)
+	if(animate)
+		FLICK("valve[!src.on][src.on]", src)
+	src.icon_state = "valve[src.on]"
+
+/obj/machinery/fluid_machinery/binary/valve/refresh_network(datum/flow_network/network)
+	..()
+	if(!src.on)
+		return
+	if(src.network1 && src.network2)
+		src.network1.merge_network(src.network2)
+
+/obj/machinery/fluid_machinery/binary/valve/disposing()
+	src.on = FALSE
+	src.network1.rebuild_network_force()
+	..()
+
+ABSTRACT_TYPE(/obj/machinery/fluid_machinery/trinary)
+/obj/machinery/fluid_machinery/trinary
+	var/datum/flow_network/network1
+	var/datum/flow_network/network2
+	var/datum/flow_network/network3
+
+/obj/machinery/fluid_machinery/trinary/New()
+	..()
+	switch(src.dir)
+		if(NORTH)
+			src.initialize_directions = NORTH|EAST|SOUTH
+		if(EAST)
+			src.initialize_directions = EAST|SOUTH|WEST
+		if(SOUTH)
+			src.initialize_directions = SOUTH|WEST|NORTH
+		if(WEST)
+			src.initialize_directions = WEST|NORTH|EAST
+
+/obj/machinery/fluid_machinery/trinary/disposing()
+	src.network1?.machines -= src
+	src.network2?.machines -= src
+	src.network3?.machines -= src
+	..()
+
+/obj/machinery/fluid_machinery/trinary/initialize()
+	for(var/obj/fluid_pipe/target in get_step(src, turn(src.dir, 180)))
+		if(target.initialize_directions & get_dir(target,src))
+			src.network1 = target.network
+			src.network1.machines += src
+			break
+
+	for(var/obj/fluid_pipe/target in get_step(src, turn(src.dir, -90)))
+		if(target.initialize_directions & get_dir(target,src))
+			src.network2 = target.network
+			src.network2.machines += src
+			break
+
+	for(var/obj/fluid_pipe/target in get_step(src, src.dir))
+		if(target.initialize_directions & get_dir(target,src))
+			src.network3 = target.network
+			src.network3.machines += src
+			break
+
+/obj/machinery/fluid_machinery/trinary/refresh_network(datum/flow_network/network)
+	src.network1?.machines -= src
+	src.network2?.machines -= src
+	src.network1 = null
+	src.network2 = null
+	src.initialize()
+
+/obj/machinery/fluid_machinery/trinary/filter
+	name = "Reagent Filter"
+	desc = "Filters out a specific reagent."
+	HELP_MESSAGE_OVERRIDE("Can be loaded with a <b>beaker</b>, which must contain at least 1 unit of a reagent. The most plentiful reagent is chosen for filtering.")
+	icon_state = "filter0"
+	flags = NOSPLASH
+	var/pullrate = 200
+	var/obj/item/reagent_containers/glass/beaker
+
+/obj/machinery/fluid_machinery/trinary/filter/attackby(obj/item/reagent_containers/glass/B, mob/user)
```

Comment:
Needs a swap mechanic for when it already has a beaker, currently just deletes the one that's in there.